feat: Preventing Spark OOM kills best practices (#1041)
awslabs/data-on-eks: website/docs/bestpractices/analytics/spark-oom-kills.md

---
sidebar_position: 4
sidebar_label: Preventing OOM Kills in Spark on EKS
---
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Preventing OOM Kills in Spark on EKS
# Preventing OOM Kills in Spark on Kubernetes
Every organization running large scale Spark workloads on Kubernetes has dealt with this: a job runs for hours, processes terabytes of data, completes 80% of its work, and then executors start disappearing. No JVM exception. No heap dump. No warning in Spark UI. Just `exit code 137` and hours of compute burned. The standard response is to throw more memory at it, bump `memoryOverhead` by another 10 GB, and hope for the best. That works until the next data spike.

> The root cause is not insufficient memory. It is a design flaw in how **cgroupsv1** handles the Linux page cache. When a Spark executor reads shuffle data from local NVMe, the kernel caches those file pages in RAM. Under cgroupsv1, this page cache counts against the container's memory limit with no mechanism to reclaim it before the OOM killer fires. The kernel kills your executor to free memory it could have simply evicted.

Reviewer: This quite a generalization. Can all OOM scenarios can be attributed to cache being counted against memory usage?

Author: not all OOM scenarios are caused by page cache accounting. OOM kills can also result from actual heap pressure, off heap memory growth, or undersized memoryOverhead as well. Will update the paragraph.

> The root cause is not insufficient memory. It is a design flaw in how **cgroupsv1** handles the Linux page cache. When a Spark executor reads shuffle data from local NVMe, the kernel caches those file pages in RAM. Under cgroupsv1, this page cache counts against the container's memory limit with no mechanism to reclaim it before the OOM killer fires. The kernel kills your executor to free memory it could have simply evicted.

Reviewer: ```suggestion
The root cause is not insufficient memory. It is a design limitation in how **cgroupsv1** handles the Linux page cache. When a Spark executor reads shuffle data from local storage, the kernel caches those file pages in RAM. Under cgroupsv1, this page cache counts against the container's memory limit with no mechanism to reclaim it before the OOM killer fires. The kernel kills your executor to free memory it could have simply evicted.
```

**cgroupsv2** fixes this with `memory.high`, a throttling boundary that forces page cache eviction before reaching the hard kill limit. Kubernetes exposes this through the **MemoryQoS** feature gate ([KEP-2570](https://github.com/kubernetes/enhancements/issues/2570)). This guide covers the kernel internals behind the problem, the cgroupsv2 solution, and the exact EKS configuration to deploy it.

```
                          What actually happens during a Spark OOM kill
  ┌─────────────────────────────────────────────────────────────────────────────────────┐
  │                                                                                     │
  │   Spark Executor (JVM)              Linux Kernel (cgroup memory controller)         │
  │   ════════════════════              ═══════════════════════════════════════          │
  │                                                                                     │
  │   1. Writes shuffle data    ──►     Dirty pages land in page cache                  │

>   │      to local NVMe                  Writeback flushes to disk                       │

Reviewer: ### 
```suggestion
  │      to local storage                  Writeback flushes to disk                       │
```

  │                                     ┌─────────────────────────────────────────┐      │
  │                             cgv1:   │  OOM killer fires. SIGKILL. exit 137.  │      │
  │                                     │  Page cache COULD have been evicted.   │      │
  │                                     │  Kernel didn't try. Executor is dead.  │      │
  │                                     └─────────────────────────────────────────┘      │
  │                                                                                     │
  │                                     ┌─────────────────────────────────────────┐      │
  │                             cgv2:   │  memory.high hit. Throttle allocs.     │      │
  │                                     │  Evict page cache. Usage drops.        │      │
  │                                     │  Executor survives. Job completes.     │      │
  │                                     └─────────────────────────────────────────┘      │
  │                                                                                     │
  └─────────────────────────────────────────────────────────────────────────────────────┘
```

> :::info Who is this guide for?

Reviewer: I'd move this up or remove it.

│   ├── Thread Stacks                 — one per thread (~1MB each)
│   ├── NIO Direct Buffers            — ByteBuffer.allocateDirect()
│   ├── Spark Off-Heap Pool           — Arrow/DataFusion native memory
│   └── Native malloc allocations     — JNI, compression codecs, etc.
│
└── File-Backed Memory (reclaimable)
    ├── Page Cache (clean)            — can be dropped instantly
    ├── Page Cache (dirty)            — must flush to disk first, then drop
    ├── Memory-mapped files           — mmap'd JARs, shared libraries
    └── tmpfs/shmem                   — backed by swap or RAM
```

:::tip
In a Spark executor, anonymous memory is bounded and predictable (JVM flags and Spark configs control it). Page cache is unbounded and invisible to Spark. It grows silently as the executor reads shuffle files, S3 buffers, and temp data. This invisible growth is what pushes containers over their cgroup limit.
:::

### How Page Cache Grows During Spark Execution

The page cache is the kernel's read/write buffer for all file I/O. Every `read()` and `write()` syscall goes through it. For Spark, this creates a silent memory consumer that grows across four phases:

> **Shuffle Write:** Executors write shuffle data to local NVMe. The `write()` syscall places data into page cache as dirty pages. The kernel's writeback daemon (`kworker/flush`) flushes dirty pages to disk asynchronously, governed by `vm.dirty_background_ratio` (default 10%) and `vm.dirty_ratio` (default 20%).

Reviewer: ```suggestion
**Shuffle Write:** Executors write shuffle data to local storage. The `write()` syscall places data into page cache as dirty pages. The kernel's writeback daemon (`kworker/flush`) flushes dirty pages to disk asynchronously, governed by `vm.dirty_background_ratio` (default 10%) and `vm.dirty_ratio` (default 20%).
```

│   ├── Shuffle Client    CometShuffleManager, SortShuffleManager
│   └── Compression       LZ4/Zstd codec native buffers
│       Scales with concurrent shuffle connections (3x data ≈ 3x buffers)
│
└── Linux Page Cache      ⚠️ NOT tracked by Spark OR the JVM
    ├── Shuffle Files      Written to NVMe, cached on read-back
    ├── Spill Files        Execution/storage pool spills to disk
    ├── S3A Disk Buffers   S3 reads cached on local filesystem
    └── JAR/Class Files    Loaded libraries and dependencies
        Grows unbounded — kernel counts this against cgroup limit
```

:::danger
NIO Direct Buffers and Linux Page Cache are invisible to Spark's memory accounting. Spark UI will show comfortable headroom while the container is actually approaching its cgroup limit. This is why OOM kills appear random when looking at Spark metrics alone.
:::

### Why cgroupsv1 Makes This Fatal

cgroupsv1 (the default on Amazon Linux 2) has one memory knob: `memory.limit_in_bytes`. The usage counter tracks the sum of all memory charged to the cgroup:

```
memory.usage_in_bytes = anonymous_memory (RSS) + file_cache (page cache) + kernel memory
```

> When usage hits the limit, the cgroup OOM killer fires immediately. No throttling, no preferential page cache reclaim, no second chance.

Reviewer: ```suggestion
When usage hits the limit, the cgroup OOM killer fires immediately, without ever giving the kernel a chance to reclaim page cache.
```
